### PR TITLE
THOR - Data Set Notifications With Counts

### PR DESCRIPTION
**Pull request by arciisine**

### CODE REVIEW

**Comment by arciisine** [comment]
Ties into https://github.com/NextCenturyCorporation/neon-server/pull/11

File: src/app/dashboard/abbreviate.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+const suffixes = ['k', 'M', 'G', 'T', 'P', 'E'];
+
+@Pipe({
+  name: 'abbreviate'
+})
+export class AbbreviatePipe implements PipeTransform {
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/dashboard/abbreviate.pipe.ts:8]: Should we move this pipe class to a separate folder (like, maybe `src/app/pipes`)?  It seems general-use.

File: src/app/dashboard/dashboard.component.ts [hunk 2]
```diff
@@ -589,8 +589,8 @@ export class DashboardComponent implements AfterViewInit, OnInit, OnDestroy {
     /**
      * Indicates to the dashboard that there is new data available
      */
-    dataAvailableDashboard() {
-        this.updatedData = true;
+    dataAvailableDashboard(event: { message: MessageEvent }) {
+        this.updatedData += (JSON.parse(event.message.data).count || 1);
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/dashboard/dashboard.component.ts:592]: @arciisine I think maybe we shouldn't assume that the count is 1 if it's not provided... Could we just show a blank red dot instead?

File: src/app/dashboard/dashboard.component.html [hunk 3]
```diff
@@ -32,9 +32,9 @@
     <mat-toolbar class="neon-toolbar-large neon-toolbar-wide titlebar-header-text" color="primary" layout-align="space-between center" flex>
         <span>{{ datasetService.getCurrentDashboardTitle() || "Choose Dashboard..." }}</span>
         <span class="fill-remaining"></span>
-        <button mat-icon-button class="neon-icon-button-large notify" aria-label="Notifications" matBadgePosition="above after"
-            [matBadgeHidden]="!updatedData" matBadgeColor="warn" matBadge="x" matBadgeSize="small" (click)="refreshDashboard()"
-            matTooltip="Refresh Data" tooltip-position="below">
+        <button mat-icon-button class="neon-icon-button-large notify" aria-label="Notifications" matBadgePosition="above before"
```
Comment (ThomasSchellenbergNextCentury) [inline src/app/dashboard/dashboard.component.html:35]: @arciisine I'm getting "Maximum call stack size exceeded" errors on this line when I press the bell button.

Comment (arciisine) [reply src/app/dashboard/dashboard.component.html:35]: I'll take a look, not sure what would be recursive there

File: src/app/dashboard/abbreviate.pipe.ts (hunk 1 above)
Comment (arciisine) [reply src/app/dashboard/abbreviate.pipe.ts:8]: I generally would put it closest to it's immediate use, and pull into a more common place when the use case arises.  Specifically to limit the amount of extra general code we are putting in the common area that may only be used by one component.

File: src/app/dashboard/dashboard.component.ts (hunk 2 above)
Comment (arciisine) [reply src/app/dashboard/dashboard.component.ts:592]: The badge doesn't like it when the text is missing.  Additionally, if I get a mixture of countable and non-countable events, how do I notify the user of changes.  I'd rather go the other way and make count a required field on notification.

File: src/app/dashboard/dashboard.component.html (hunk 3 above)
Comment (arciisine) [reply src/app/dashboard/dashboard.component.html:35]: The button was triggering a refresh event, and refresh events were being listened to by the dashboard, and calling the same method (hence recursion).  We should be good now.

**Review by ThomasSchellenbergNextCentury — APPROVED** [review]

**Pull request merged**